### PR TITLE
ARROW-16898: [Python] Fix pandas conversion failure when using non-str index name

### PR DESCRIPTION
**Pull request by martin-liu**
Ticket: https://issues.apache.org/jira/browse/ARROW-16898 

When do `Table.from_pandas(df)`, current code didn't convert `index` name to str (it did [convert `column` name to str](https://github.com/apache/arrow/blob/apache-arrow-8.0.0/python/pyarrow/pandas_compat.py#L356)), so that it will fail if **non-str index name** in df.

Code to reproduce:
```python
import pandas as pd
import pyarrow as pa

df = pd.DataFrame({0: [1, 2, 3], 1: [4, 5, 6]})
df = df.set_index(0)
pa.Table.from_pandas(df)
```

Error:
```python
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [3], in <module>
      4 df = pd.DataFrame({0: [1, 2, 3], 1: [4, 5, 6]})
      5 df = df.set_index(0)
----> 6 pa.Table.from_pandas(df)

File ~/src/mlpsandboxrt/venv/lib/python3.8/site-packages/pyarrow/table.pxi:1394, in pyarrow.lib.Table.from_pandas()

File ~/src/mlpsandboxrt/venv/lib/python3.8/site-packages/pyarrow/pandas_compat.py:610, in dataframe_to_arrays(df, schema, preserve_index, nthreads, columns, safe)
    608     for name, type_ in zip(all_names, types):
    609         name = name if name is not None else 'None'
--> 610         fields.append(pa.field(name, type_))
    611     schema = pa.schema(fields)
    613 pandas_metadata = construct_metadata(df, column_names, index_columns,
    614                                      index_descriptors, preserve_index,
    615                                      types)

File ~/src/mlpsandboxrt/venv/lib/python3.8/site-packages/pyarrow/types.pxi:1698, in pyarrow.lib.field()

File stringsource:15, in string.from_py.__pyx_convert_string_from_py_std__in_string()

TypeError: expected bytes, int found
```

This PR uses `_column_name_to_strings` to convert the index name to str before use it.

### CODE REVIEW

**Comment by lidavidm** [comment]
I think that's been happening in several PRs (I don't think an issue has been filed yet…)

**Comment by lidavidm** [comment]
Filed ARROW-16873

I'll try to review this tomorrow, but a quick question, such index/column names won't round-trip back to Pandas right? Should we note that somewhere in the docs?

**Comment by martin-liu** [comment]
@lidavidm Thanks for your time! 
I didn't fully get your point, basically Pandas allows non-str column/index names, it's just in PyArrow we are using `str(name)`. 
The existing code has the str conversion for `columns`, but missing for `indexes`, so I added it in this PR.

What kind of doc change are you thinking?

**Comment by lidavidm** [comment]
Actually, it turns out that integer column names are preserved:

```python
>>> pa.Table.from_pandas(pd.DataFrame({'123': [1]})).to_pandas().columns
Index(['123'], dtype='object')
>>> pa.Table.from_pandas(pd.DataFrame({123: [1]})).to_pandas().columns
Int64Index([123], dtype='int64')
```

Is this also the case here?

File: python/pyarrow/tests/test_pandas.py [hunk 1]
```diff
@@ -149,6 +149,12 @@ def test_non_string_columns(self):
         table = pa.Table.from_pandas(df)
         assert table.field(0).name == '0'
 
+    def test_non_string_columns_with_index(self):
+        df = pd.DataFrame({0: [1, 2, 3], 1: [4, 5, 6]})
+        df = df.set_index(0)
+        table = pa.Table.from_pandas(df)
+        assert table.field(0).name == '1'
```
Comment (lidavidm) [inline python/pyarrow/tests/test_pandas.py:156]: I suppose, assuming round-tripping the type of the index names works, we should also `_check_pandas_roundtrip(df, preserve_index=True)` here

Comment (martin-liu) [reply python/pyarrow/tests/test_pandas.py:156]: Good point! I tested that `_check_pandas_roundtrip` will fail

**Comment by lidavidm** [comment]
I think it doesn't work for mixed type, you should probably see a warning?

Regardless that was my original point - if it doesn't fully round-trip it's not _necessarily_ a problem IMO (CC @jorisvandenbossche) but it should be noted (though if we can get it to round-trip like everything else that would be great)

**Comment by lidavidm** [comment]
Looks fine by me (though maybe we want this warning to only ever fire once, instead of once per level/once per call?) but I'll let someone else also take a look.

Also, at this point, can we file a Jira for this? It's no longer a minor patch.

**Comment by martin-liu** [comment]
@lidavidm Created ticket https://issues.apache.org/jira/browse/ARROW-16898 for this, also adjust the warning out from the loop.

**Review by lidavidm — APPROVED** [review]

**Comment by lidavidm** [comment]
Thank you @martin-liu!

**Comment by martin-liu** [comment]
@lidavidm I appreciate your quick response and helpful advices, thank you very much!

**Pull request merged**